### PR TITLE
Fix group payment and attendee kick-in levels

### LINKED ISSUE #1257: Group payment should check all group members
The ``/preregistation/group_members`` page has a payment link which appears if the group is unpaid.  The two most common things which would cause this are
- the group has unpaid badges
- the group has one or more members who kicked in extra but then didn't pay

There's code in ``/preregistration/process_group_payment`` which checks the second condition - but only for the group leader.  It should check it for all group members instead of only the group leader.

### LINKED ISSUE #1584: Group leader with kick-in doesn't show kick-in $$ paid on attendee screen
This screen says paid by group, $0 owed, $0 paid, $### kick-in chosen.  This leaves the reader feeling like the kick-in hasn't been paid for.

### PR DESCRIPTION
**Pull request by kitsuta**
Fixes #1584. While working on it I fixed #1257 which had apparently not been fixed even though I thought it was.

### CODE REVIEW

File: uber/site_sections/preregistration.py [hunk 1]
```diff
@@ -339,13 +339,12 @@ def process_group_payment(self, session, payment_id, stripeToken):
         else:
             group.amount_paid += charge.dollar_amount
 
-            for attendee in charge.attendees:
+            for attendee in group.attendees:
```
Comment (kitsuta) [inline uber/site_sections/preregistration.py]: @EliAndrewC - I would love an explanation for _why_ this change fixes #1584, because frankly, I'm a bit lost.

Comment (EliAndrewC) [inline uber/site_sections/preregistration.py]: The ``uber.utils.Charge`` class takes a list of objects which can be either ``Attendee`` or ``Group`` objects.  If you say ``charge.attendees`` then it will return a list of ``Attendee`` objects which were explicitly passed to the ``Charge`` class.

Whereas if you have a ``group`` which you got out of the ``charge`` and then say ``group.attendees`` then you'll get all attendees which are part of that group, even if they were not explicitly part of the ``Charge`` object.  Since we get these from the database via the foreign keys into the group class, we don't need to say ``session.merge``.

The ``Charge`` class could definitely use some documentation :(

Comment (kitsuta) [inline uber/site_sections/preregistration.py]: Well, that is good to know, but it doesn't quite explain why session.merge _wasn't working_ before. It would clearly find the attendees (because the group's `paid` value would be correct) but the attendees' `paid` property was never updated.

**Comment by kitsuta** [comment]
I wasn't confident in fixing the merge conflicts back when GitHub's method automatically merged the PR when you did that, but now they're fixed. This should be re-reviewed though, just to make sure it won't mess anything up -- I'm not totally confident in these changes.

File: uber/site_sections/preregistration.py [hunk 2]
```diff
@@ -260,7 +260,7 @@ def dealer_confirmation(self, session, id):
     @log_pageview
     def group_members(self, session, id, message=''):
         group = session.group(id)
-        charge = Charge([group, group.leader]) if group.leader else Charge(group)
+        charge = Charge([group] + listify(group.attendees))
```
Comment (RobRuana) [inline uber/site_sections/preregistration.py]: Is this change required? If we use `group.attendees` below, then we shouldn't need to add each attendee as a target for the `Charge`, correct?

Comment (kitsuta) [inline uber/site_sections/preregistration.py]: Hm, I think you may be right. I'll look into it!

**Comment by kitsuta** [comment]
OK, after testing further, it looks like what ended up happening is that we decided to go the other way on this workflow. That is, individual members' kick-in levels are not included on the group payment screen. That means there's no reason to change the code -- it is already consistent and functional.

File: uber/site_sections/preregistration.py (hunk 2 above)
Comment (kitsuta) [inline uber/site_sections/preregistration.py]: OK, so the original intention of this was to actually charge people for attendees' unpaid kick-in levels. This `charge` value gets passed to the stripe form on the page itself -- the other part of the code is what we mark the attendees using. Getting rid of this would normally have made us mark attendees as paid but not actually charge for their unpaid kick-in levels.

However, since this PR was submitted, I changed a lot about how group payments work. I'm about to update the branch with the new code and I'll explain it in full.

**Comment by kitsuta** [comment]
OK, after even further review (and a bit of my memory being jogged), I realized this should indeed be merged.

CURRENT BEHAVIOR:
1. Someone registers a group. At this time, all PAID_BY_GROUP attendees' extras are included in group payment, and they are marked individually as having paid for those extras. Functionally, this means that the group leader is paying for their own extras in the same transaction as pre-registering and paying for their group, which makes perfect sense.
2. At any time, a group leader can update their extras. The system will ask them to pay immediately, though they can simply close the page to not pay. If the group has been paid for, there is no Stripe button on the group management page (due to an if-check in the template).
3. HOWEVER, if the group has `amount_unpaid`, a Stripe button will show up on the group management page. Furthermore -- and this is the crux of bug #1584 -- anyone paying through that button _will also pay for the group leader's extras._ Even worse, the group leader is never marked as having paid for their own extras when you pay through this Stripe button. This situation is rare and a bit convoluted, but it can still happen, and it's a bit scary because it means we can take people's money and not record it.

THIS PR:
1. Exact same as current behavior.
2. No matter what happens, after the group exists in the database, attendees' extras are not included in the group payment. Therefore, the `process_group_payment` block (which is called from the group management page) need not worry about the group's attendees' extras.
3. If a Stripe button exists on the group management page, it will only pay for group members' badge costs (which is what we want).

As a point of interest, if we ever want to change how we count attendees' extras as group extras, this is the block of code I wrote to account for that in `process_group_payment` before I realized it was unnecessary:
```python
            # For new groups, we add paid_by_group attendees' total extras to the group's amount_extra.
            # Therefore, we want to mark attendees as having paid for their extras when the payment is processed.
            if group.amount_extra:
                group.amount_extra = 0
                for attendee in group.attendees:
                    if attendee.paid == c.PAID_BY_GROUP:
                        attendee.amount_paid += attendee.total_cost - attendee.badge_cost
                        session.merge(attendee)
```

**Comment by kitsuta** [comment]
Interesting to note: Because of the way Charge handles models, it seems that a group taken from the Charge class will always have a truthy `is_new` property even if the real group is not new. I've updated this PR to get around that, but it may bite us in the future.

File: uber/site_sections/preregistration.py [hunk 3]
```diff
@@ -260,7 +260,7 @@ def dealer_confirmation(self, session, id):
     @log_pageview
     def group_members(self, session, id, message=''):
         group = session.group(id)
-        charge = Charge([group, group.leader]) if group.leader else Charge(group)
+        charge = Charge(group)
```
Comment (RobRuana) [inline uber/site_sections/preregistration.py]: Ok, I think we still want to explicitly add `group.leader` as one of the `targets` for `Charge`, because `group.attendees` will not necessarily include the `leader`.

But I am out of my depth when it regarding what is _supposed to happen_. I am not a domain expert here :(

Comment (kitsuta) [inline uber/site_sections/preregistration.py]: This is a page where we explicitly _don't_ want to charge for the attendees' badges. Groups have their own properties/columns for money.

Comment (RobRuana) [inline uber/site_sections/preregistration.py]: Okay, clearly out of my depth :)

Comment (kitsuta) [inline uber/site_sections/preregistration.py]: To be fair, it took me a few tries to remember how this works and I wrote it!
